### PR TITLE
Ignore error when discovering server supported groupVersions and resources

### PR DESCRIPTION
**Pull request by caesarxuchao**
First two commits are #15973.

Let the Discovery Client ignore error when discovering server supported groupVersions

fixes @bgrant0607's comment https://github.com/kubernetes/kubernetes/pull/15796#discussion_r42576181.

This one needs to be cherry-picked.

supersede #16066

cc @liggitt @jlowdermilk

### CODE REVIEW

**Comment by liggitt** [comment]
cc @deads2k for negotiation awareness

File: pkg/apis/unversioned/codec/json_codec.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type JSONCodec struct{}
```
Comment (liggitt) [inline pkg/apis/unversioned/codec/json_codec.go]: I don't think we want this... why can't we use an existing codec for this? Any of them should be able to handle unversioned objects, right?

File: pkg/client/unversioned/discovery_client.go [hunk 2]
```diff
@@ -71,42 +68,21 @@ func apiVersionsToAPIGroup(apiVersions *unversioned.APIVersions) (apiGroup unver
...
 	// Get the groupVersions exposed at /api
-	url := d.baseURL
-	url.Path = "/api"
-	resp, err := d.get(url.String())
-	if err != nil {
-		return nil, err
+	v := &unversioned.APIVersions{}
+	err = d.Get().AbsPath("/api").Do().Into(v)
+	apiGroup := unversioned.APIGroup{}
+	if err == nil {
```
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: Not all errors are ignorable.  You probably only want to ignore 404's, right?

File: pkg/client/unversioned/discovery_client.go [hunk 3]
```diff
@@ -71,42 +68,21 @@ func apiVersionsToAPIGroup(apiVersions *unversioned.APIVersions) (apiGroup unver
...
-	resp2, err := d.get(url.String())
-	if err != nil {
-		return nil, err
-	}
-	defer resp2.Body.Close()
 	apiGroupList = &unversioned.APIGroupList{}
-	if err = json.NewDecoder(resp2.Body).Decode(&apiGroupList); err != nil {
-		return nil, fmt.Errorf("unexpected error: %v", err)
+	err = d.Get().AbsPath("/apis").Do().Into(apiGroupList)
+	if err != nil {
```
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: Why suppress all errors here?  It seems like most errors are worth propagating.

File: pkg/client/unversioned/discovery_client.go [hunk 4]
```diff
@@ -116,21 +92,17 @@ func (d *DiscoveryClient) ServerGroups() (apiGroupList *unversioned.APIGroupList
 
 // ServerResourcesForGroupVersion returns the supported resources for a group and version.
 func (d *DiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (resources *unversioned.APIResourceList, err error) {
-	url := d.baseURL
+	url := url.URL{}
 	if groupVersion == "v1" {
```
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: Code like this is worrying me.  Is the next version of the kube API going to have a group?

File: pkg/client/unversioned/discovery_client.go [hunk 5]
```diff
@@ -154,7 +126,9 @@ func (d *DiscoveryClient) ServerResources() (map[string]*unversioned.APIResource
 
 func setDiscoveryDefaults(config *Config) error {
 	config.Prefix = ""
-	config.Version = ""
+	config.Version = "unversioned"
```
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: This surprises me.  When "unversioned" become special cased?

File: pkg/client/unversioned/discovery_client.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: Yes. I will fix it.

File: pkg/client/unversioned/discovery_client.go (hunk 4 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: I don't follow your question, could you clarify?

File: pkg/client/unversioned/discovery_client.go (hunk 5 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: The `Version` is not being used. I put it there to pass the check [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/helper.go#L392). Is it ok?

File: pkg/client/unversioned/discovery_client.go (hunk 4 above)
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: > I don't follow your question, could you clarify?

When kube gets a `v2`, will this code have to become?
```go
if groupVersion == "v1" || groupVersion == "v2"{
```

File: pkg/apis/unversioned/codec/json_codec.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/apis/unversioned/codec/json_codec.go]: If we use the Scheme codec, I think it would probably fail [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/decode.go#L106) and [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/decode.go#L123).

Even if the Scheme can handle the unversioned objects, I feel it's less robust to use such a complex codec when we only need a simple one. What do you think?

File: pkg/client/unversioned/discovery_client.go (hunk 4 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: No. I think the plan is breaking the kube API to smaller groups and each group evolves in their own pace. We won't have a monolithic v2 API.

File: pkg/client/unversioned/discovery_client.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: @deads2k this is fixed. PTAL. Thank you.

File: pkg/client/unversioned/discovery_client.go (hunk 3 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: Fixed.

File: pkg/client/unversioned/discovery_client.go [hunk 6]
```diff
@@ -71,42 +69,29 @@ func apiVersionsToAPIGroup(apiVersions *unversioned.APIVersions) (apiGroup unver
...
-	if err != nil {
-		return nil, err
-	}
-	defer resp2.Body.Close()
 	apiGroupList = &unversioned.APIGroupList{}
-	if err = json.NewDecoder(resp2.Body).Decode(&apiGroupList); err != nil {
-		return nil, fmt.Errorf("unexpected error: %v", err)
+	err = d.Get().AbsPath("/apis").Do().Into(apiGroupList)
+	if err != nil {
+		if !errors.IsNotFound(err) {
```
Comment (liggitt) [inline pkg/client/unversioned/discovery_client.go]: A secured 1.0 server could easily return a forbidden error here

File: pkg/client/unversioned/discovery_client.go (hunk 4 above)
Comment (liggitt) [inline pkg/client/unversioned/discovery_client.go]: Does `/api/v1` return a list of server resources in 1.0? Is this expected to work against 1.0 servers?

File: pkg/client/unversioned/discovery_client.go (hunk 5 above)
Comment (liggitt) [inline pkg/client/unversioned/discovery_client.go]: That seems not good. Setting bogus values just to pass checks in other parts of the code…

File: pkg/client/unversioned/discovery_client.go (hunk 4 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: No, that won't work. I will fix it in another PR.

File: pkg/client/unversioned/discovery_client.go (hunk 6 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: I'm not familiar with the `secured server`. Could you point me to the related code? Thanks.

File: pkg/apis/unversioned/codec/json_codec.go (hunk 1 above)
Comment (lavalamp) [inline pkg/apis/unversioned/codec/json_codec.go]: I don't have a strong opinion. If we find ourselves adding anything more complicated than straight json calls here, we should consider not using this.

Comment (lavalamp) [inline pkg/apis/unversioned/codec/json_codec.go]: But consider changing the name to jsonCodec, and making a `func NewCodec() jsonCodec {...}` so people can make one? Or a public variable `var Codec = jsonCodec{}`.

File: pkg/client/unversioned/discovery_client.go (hunk 4 above)
Comment (lavalamp) [inline pkg/client/unversioned/discovery_client.go]: Yes, "v1" is the only exception to the group/version format we ever intend to have.

File: pkg/client/unversioned/discovery_client.go [hunk 7]
```diff
@@ -116,21 +101,17 @@ func (d *DiscoveryClient) ServerGroups() (apiGroupList *unversioned.APIGroupList
...
-	url := d.baseURL
+	url := url.URL{}
 	if groupVersion == "v1" {
 		url.Path = "/api/" + groupVersion
 	} else {
 		url.Path = "/apis/" + groupVersion
 	}
-	resp, err := d.get(url.String())
+	resources = &unversioned.APIResourceList{}
+	err = d.Get().AbsPath(url.String()).Do().Into(resources)
```
Comment (lavalamp) [inline pkg/client/unversioned/discovery_client.go]: If an error happens, does this get our nice status error type? If not that'd be a reason to switch to a more complicated codec.

File: pkg/client/unversioned/discovery_client.go (hunk 6 above)
Comment (liggitt) [inline pkg/client/unversioned/discovery_client.go]: A server that limits users access to URLs outside their namespace or specifically whitelist end exceptions (for 1.0, that would likely include `/api`, but definitely wouldn't include `/apis`. This is close to possible with the abac authorizer (https://github.com/kubernetes/kubernetes/issues/13097 is open to make it work completely), but real installations would be likely to deny requests outside the expected URL paths. TL;DR: either a NotFound or Forbidden error could easily be received from a 1.0 installation.

File: pkg/client/unversioned/discovery_client.go (hunk 5 above)
Comment (lavalamp) [inline pkg/client/unversioned/discovery_client.go]: Consider making another RESTClient constructor instead of this.

Comment (lavalamp) [inline pkg/client/unversioned/discovery_client.go]: Actually, maybe it's OK to clean this up later.

File: pkg/client/unversioned/discovery_client.go (hunk 3 above)
Comment (lavalamp) [inline pkg/client/unversioned/discovery_client.go]: Fix looks good.

File: pkg/client/unversioned/discovery_client.go (hunk 7 above)
Comment (lavalamp) [inline pkg/client/unversioned/discovery_client.go]: Consider eating NotFound errors here, too?

**Comment by lavalamp** [comment]
Only minor comments, thanks.

File: pkg/client/unversioned/discovery_client.go (hunk 6 above)
Comment (liggitt) [inline pkg/client/unversioned/discovery_client.go]: @deads2k  I would actually be very tolerant of errors when requesting `/apis` if we got it usable response from `/api`, if we ever expect this to work against a 1.0 server

File: pkg/client/unversioned/discovery_client.go (hunk 7 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: We have to. But I prefer to make that another PR.

Comment (liggitt) [inline pkg/client/unversioned/discovery_client.go]: > I prefer to make that another PR

If it's a small change that's easier to understand in the context of this PR, let's go ahead and fold it in

Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: >If an error happens, does this get our nice status error type?

Yes.

v1.1 server does not encode Status, so the unmarshal the json will decode it.

in v1.0, Status is in v1 types.go, server will encode Status to v1. jsonCodec will decode it to unversioned.Status and ignore the TypeMeta. So it still works.

Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: @liggitt ok :)

File: pkg/client/unversioned/discovery_client.go (hunk 5 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: I'll leave a todo and clean it up later. I believe we'll make more changes to RESTClient in the near future.

File: pkg/client/unversioned/discovery_client.go (hunk 6 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: Now it ignores 403 error as well.

File: pkg/client/unversioned/discovery_client.go (hunk 7 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: added.

File: pkg/apis/unversioned/codec/json_codec.go (hunk 1 above)
Comment (caesarxuchao) [inline pkg/apis/unversioned/codec/json_codec.go]: fixed.

**Comment by caesarxuchao** [comment]
e2e failure:
```
Cluster failed to initialize within 300 seconds.
```

**Comment by caesarxuchao** [comment]
@k8s-bot test this please.

**Comment by caesarxuchao** [comment]
@k8s-bot test this please.

**Comment by caesarxuchao** [comment]
@liggitt @deads2k could you help review the entire change (including the first 2 commits which are  #15973)? We need to cherry-pick this into 1.1 release and lavalamp is not available today. Thanks.

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by caesarxuchao** [comment]
@k8s-bot test this

**Comment by jlowdermilk** [comment]
@caesarxuchao the bot is currently dead, we're trying to revive it #16161.

**Comment by caesarxuchao** [comment]
@jlowdermilk Thanks! I'm just going to reach out for help :)

File: pkg/apis/unversioned/codec/json_codec.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type jsonCodec struct{}
```
Comment (deads2k) [inline pkg/apis/unversioned/codec/json_codec.go]: This is still strange.  Why can't we use one of our normal codecs?

File: pkg/client/unversioned/discovery_client.go [hunk 9]
```diff
@@ -71,42 +69,29 @@ func apiVersionsToAPIGroup(apiVersions *unversioned.APIVersions) (apiGroup unver
...
+	apiGroup := unversioned.APIGroup{}
+	if err == nil {
+		apiGroup = apiVersionsToAPIGroup(v)
 	}
-	var v unversioned.APIVersions
-	defer resp.Body.Close()
-	err = json.NewDecoder(resp.Body).Decode(&v)
-	if err != nil {
-		return nil, fmt.Errorf("unexpected error: %v", err)
+	if err != nil && !errors.IsNotFound(err) && !errors.IsForbidden(err) {
```
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: move this before the the `if err == nil` and you won't need to conditionally set `apiGroup`

Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: > move this before the the if err == nil and you won't need to conditionally set apiGroup

nm, that was wrong.

File: pkg/client/unversioned/discovery_client.go (hunk 3 above)
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: Can you make this structure parallel with checks above in the same function.

File: pkg/client/unversioned/discovery_client.go [hunk 10]
```diff
@@ -154,7 +140,11 @@ func (d *DiscoveryClient) ServerResources() (map[string]*unversioned.APIResource
 
 func setDiscoveryDefaults(config *Config) error {
 	config.Prefix = ""
-	config.Version = ""
+	// "unversioned" is a bogus value to pass the check in RESTClientFor.
+	// TODO: make a RESTClient constructor that accepts empty version.
+	config.Version = "unversioned"
```
Comment (deads2k) [inline pkg/client/unversioned/discovery_client.go]: Definitely not "unversioned", that's very confusing.  Also, doing this builds a `RESTClient` that has an invalid `baseURL` which disturbs me.

If you need to be able specify an empty version, I would rather see you refactor the code you're calling.  Even if it means another method to do it cleanly/easily.

**Comment by deads2k** [comment]
The fake `.Version` bothers me because of the not really valid object returned.  Have our own special codec also bothers me.  I'd like to know why our existing ones don't work.

File: pkg/client/unversioned/discovery_client.go (hunk 10 above)
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: Ok, I'll make another method.

**Comment by caesarxuchao** [comment]
In another reply to liggitt,
"If we use the Scheme codec, I think it would probably fail [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/decode.go#L106) and [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/decode.go#L123)."

I'll quickly try if the Sheme codec will fail.

**Comment by deads2k** [comment]
@caesarxuchao why would I want to return a non-runtime.Object?

**Comment by caesarxuchao** [comment]
> @caesarxuchao why would I want to return a non-runtime.Object?

sorry I don't follow. Which code are you talking about?

**Comment by deads2k** [comment]
> sorry I don't follow. Which code are you talking about?

Codecs should be able to handle `runtime.Objects` which are expected to contain the fields from `TypeMeta`.  I don't see why I wouldn't want this server-side API to return an object embeding `TypeMeta`.  It should be backwards compatible right?  Extra fields are ignored during decode.

That would fix us moving forward and contain the direct json decode to the legacy cases, right?

**Comment by caesarxuchao** [comment]
All those unversioned types doesn't include a TypeMeta:

https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/types.go#L284

I think that's a design choice we made when introducing the discovery mechanism.

**Comment by deads2k** [comment]
> All those unversioned types doesn't include a TypeMeta:
> 
> https://github.com/kubernetes/kubernetes/blob/master/pkg/api/unversioned/types.go#L284
> 
> I think that's a design choice we made when introducing the discovery mechanism.

Do you have a link to the original discussion?

@lavalamp @smarterclayton @kubernetes/goog-csi Why would we want to have our server return structured json that does not include `TypeMeta` information for decoding client-side?  It seems more like an oversight than a design objective.

**Comment by smarterclayton** [comment]
We would have to define what unversioned meant.  And we'd have to reserve
kind names (neither of which I'm opposed to)?

**Comment by deads2k** [comment]
> We would have to define what unversioned meant.  And we'd have to reserve
> kind names (neither of which I'm opposed to)?

This is already handled for `Status` isn't it?

**Comment by caesarxuchao** [comment]
@deads2k 

> Do you have a link to the original discussion?

https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37591134

There are other discussion which I can't find..

**Comment by caesarxuchao** [comment]
@deads2k 
> "If we use the Scheme codec, I think it would probably fail [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/decode.go#L106) and [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/conversion/decode.go#L123)."

Just tried with a v1.Codec, and it does fail as expected
```
unexpected error: no kind is registered for the type unversioned.APIVersions
```

File: pkg/client/unversioned/helper.go [hunk 11]
```diff
@@ -414,6 +414,31 @@ func RESTClientFor(config *Config) (*RESTClient, error) {
 	return client, nil
 }
 
+// UnversionedRESTClientFor is the same as RESTClientFor, except that it allows
+// the config.Version to be empty.
+func UnversionedRESTClientFor(config *Config) (*RESTClient, error) {
```
Comment (caesarxuchao) [inline pkg/client/unversioned/helper.go]: @deads2k 
I forked another constructor that doesn't do the empty version check.

File: pkg/client/unversioned/discovery_client.go [hunk 12]
```diff
@@ -155,6 +141,8 @@ func (d *DiscoveryClient) ServerResources() (map[string]*unversioned.APIResource
 func setDiscoveryDefaults(config *Config) error {
 	config.Prefix = ""
 	config.Version = ""
```
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: @deads2k Version is set to empty. The base URL will be to the host root.

**Comment by caesarxuchao** [comment]
@deads2k I address your comments about config.Version="unversioned", PTAL. Thank you.

Regarding the jsonCodec, we need it unless we add TypeMeta to the types in unversioned.go (which we can have another debate :) ). And for the record, @lavalamp is ok with it (https://github.com/kubernetes/kubernetes/pull/16082#discussion_r42820807)

File: pkg/api/unversioned/register.go [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unversioned
+
+func (*APIVersions) IsAnAPIObject()     {}
```
Comment (deads2k) [inline pkg/api/unversioned/register.go]: I'm pretty sure that we don't want to have `runtime.Objects` that don't contain `TypeMeta`. 

@smarterclayton @lavalamp Can you confirm or deny?

**Comment by deads2k** [comment]
The `TypeMeta` thing still bothers me, but I don't think clients are any worse off.  There's one interface change here https://github.com/kubernetes/kubernetes/pull/16082#discussion_r42983006 that I think someone else should acknowledge (it bothers me enough that I would want it fixed), but if they come down on the side of merge anyway the rest of the code looks reasonable.

**Comment by deads2k** [comment]
I think we should go ahead and add the `TypeMeta` to those types and special case the old servers.

@liggitt thoughts?

**Comment by liggitt** [comment]
Any JSON we want the server to respond with should include TypeMeta. Handling the `/api` response from 1.0 servers which do not include kind/apiVersion fields should be special-cased and handled as part of the 1.0 compatibility work that still needs doing. I don't want to introduce codecs that do things that don't make sense.

**Comment by caesarxuchao** [comment]
IMO, the `Version` part of TypeMeta doesn't make sense for discovery API objects. What version should the server encode the discovery API objects to? What if the client doesn't understand the encoding version? That's why we made the discovery API objects unversioned in the first place.

I think we had considered only setting the `Kind` field, and leaving the `Version`blank, but I don't remember why we abandoned that idea.

@deads2k @liggitt, 
Do you want to add the TypeMeta field to the discovery API Objects or ***all*** objects in the unversioned/types.go?
What exactly are the benefits of adding the TypeMeta field?

**Comment by liggitt** [comment]
Unversioned does not mean unstructured, it means we are declaring we'll never need to convert between versions of these structs.

> What exactly are the benefits of adding the TypeMeta field?

The same benefits adding it to the `Status` object:
1. Identifies the target struct so a codec can deserialize correctly
2. Prevents decoding garbage JSON as an empty struct of a different kind (for example, we wouldn't want `{"foo":"bar"}` to successfully deserialize to a `Status` object
3. Allows the codec to implement `Decode` (which your `JSONDecoder` cannot, because it has nothing to use to recognize the target kind)

**Comment by smarterclayton** [comment]
> I'm pretty sure that we don't want to have runtime.Objects that don't
contain TypeMeta.

There has to be a really, really good reason not to.  Every runtime.Object
should be able to answer Kind.  I think it's reasonable for a
runtime.Object to answer Version = "" (which means unversioned).

**Comment by caesarxuchao** [comment]
@liggitt, Thanks for the explanation. So are you in favor of "setting the `Kind` field, and leaving the `Version` blank" for the objects in the unversioned/types.go?

**Comment by caesarxuchao** [comment]
On second thought, it might be confusing if we set the `Version` field blank for the unversioned objects, since that might be confused with the internal objects.

**Comment by smarterclayton** [comment]
Internal objects *are* unversioned.  That's why they are internal.  It's
not wrong for us to return versioned objects to unversioned APIs - it would
just be wrong for a client to assume that the fact that the version means
something (which is a reason not to return a version) or to trick a
developer into thinking they can get away with changing the version when
they mutate the object.

File: pkg/client/unversioned/helper.go (hunk 11 above)
Comment (lavalamp) [inline pkg/client/unversioned/helper.go]: cc @krousey another thing to fix by splitting layers...

File: pkg/api/unversioned/register.go (hunk 13 above)
Comment (lavalamp) [inline pkg/api/unversioned/register.go]: @deads2k So there are some uses of `runtime.Object`s that will expect to be able to find a TypeMeta. However, they shouldn't crash if they don't find it, and they shouldn't end up operating on these particular types.

Later we can reevaluate this, but at least for the moment we don't want these types to have the apiVersion/kind fields, but they *are* (unversioned) api types.

File: pkg/apis/unversioned/codec/json_codec.go (hunk 8 above)
Comment (lavalamp) [inline pkg/apis/unversioned/codec/json_codec.go]: @caesarxuchao did you try just using the normal codec? what happens?

(but I don't object to making this codec. maybe 'unversionedJSON' would be a better name for it? This is a nit though.)

Comment (liggitt) [inline pkg/apis/unversioned/codec/json_codec.go]: the codec needs at least `kind` in the incoming JSON to implement the majority of the decode functions

Comment (caesarxuchao) [inline pkg/apis/unversioned/codec/json_codec.go]: i tried with a v1 codec and it has error:
https://github.com/kubernetes/kubernetes/pull/16082#issuecomment-150677686

**Comment by caesarxuchao** [comment]
@lavalamp may I get the LGTM label?

I'll open an issue regarding adding TypeMeta to the unversioned objects. If we decide to add TypeMeta, we can do it after 1.1, and it will be ignored by 1.1 client, so it's backward compatible.

**Comment by liggitt** [comment]
Can we please not ship 1.1 with an untyped APIversions response? Right now only 1.0 /api requires special handling, I would like to keep it that way if at all possible

**Comment by caesarxuchao** [comment]
@liggitt, I understand your concern, but the "untyped APIversions response" is already merged into the 1.1 release (#13955), this PR only makes the client correctly handle the untyped response..

**Comment by liggitt** [comment]
That was an oversight which we have an opportunity to fix before requiring special handling of both /api and /apis in perpetuity.

**Comment by lavalamp** [comment]
@liggitt the responses are already unversioned. Are you objecting to that or to making them `runtime.Object`s?

**Comment by lavalamp** [comment]
And they *have* to be unversioned, or at least a version that all clients everywhere can understand.

**Comment by lavalamp** [comment]
@liggitt what special handling?

**Comment by liggitt** [comment]
Unversioned != untyped. They should have a kind so a codec can read them and return an object, just like status

**Comment by lavalamp** [comment]
@liggitt I can get behind adding a kind. But that doesn't block this PR?

**Comment by liggitt** [comment]
> what special handling?

JSONCodec

File: pkg/apis/unversioned/codec/json_codec.go (hunk 8 above)
Comment (lavalamp) [inline pkg/apis/unversioned/codec/json_codec.go]: @liggitt Those methods are not called for these objects--they are not even implemented for this codec.

@caesarxuchao I think it is a good idea to make the unversioned objects more "real".

@liggitt / @deads2k Will you be satisfied with TODO items here?

**Comment by caesarxuchao** [comment]
e2e is a flake:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pods.go:862
Oct 26 14:47:28.674: back-off gap is not increasing exponentially pod=pod-back-off-exponentially/back-off delay1=1m3.000000001s delay2=57s
```

**Comment by caesarxuchao** [comment]
@k8s-bot test this please

**Comment by lavalamp** [comment]
LGTM @liggit I think we can and should make unversioned types more real, but I also don't think that blocks this PR. Let me know if you disagree, though.

**Comment by liggitt** [comment]
> @liggitt / @deads2k Will you be satisfied with TODO items here?

The kind coming back in the APIVersions response is the primary thing I care about for 1.1. It's really hard to fix an unversioned object in "the next version"

We can clean up this path as a follow up much more easily, if we have the right data to work with.

**Comment by caesarxuchao** [comment]
add back the lgtm label because it's a squash.

**Comment by lavalamp** [comment]
@liggitt but there was never a Kind on resource listing afaik?

**Comment by smarterclayton** [comment]
Wasn't, but I think there should be as well.  It's the same argument as
status, we were just lazy about enforcing it

On Oct 26, 2015, at 6:56 PM, Daniel Smith <notifications@github.com> wrote:

@liggitt <https://github.com/liggitt> but there was never a Kind on
resource listing afaik?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16082#issuecomment-151308129>
.

**Comment by lavalamp** [comment]
Yeah, it makes sense, I just think it's a consistency problem and not a
correctness problem at this point.

**Comment by liggitt** [comment]
> Yeah, it makes sense, I just think it's a consistency problem and not a correctness problem at this point.

I disagree. The fact that we're declaring these to be runtime.Objects, and trying to read them with a client (which is good, in my opinion), yet a normal codec can't read them, speaks to the correctness of the serialization.

**Comment by deads2k** [comment]
> Yeah, it makes sense, I just think it's a consistency problem and not a
> correctness problem at this point.

`kind` provides a reliable way for clients to parse the objects.  "What kind of object is sitting in this file?  Check the kind field".  Losing the `kind` makes life harder.  Our codec uses it for a reason.  I'd expect other clients to have similar logic.

**Comment by lavalamp** [comment]
@liggitt & @deads2k These objects are not PUT/POSTable, so they won't appear in user config. Basically I'm claiming that there's no circumstance, practically speaking, where you'll read an arbitrary object and it'll turn out to be one of these, so this isn't a super important property for these objects.

**Comment by deads2k** [comment]
> @liggitt & @deads2k These objects are not PUT/POSTable, so they won't appear in user config. Basically I'm claiming that there's no circumstance, practically speaking, where you'll read an arbitrary object and it'll turn out to be one of these, so this isn't a super important property for these objects.

I could see myself in caching these locally in the near future to avoid unnecessary negotiation.  After I serialize it onto disk, I'll want to read it.  Caching it locally allows me to use the normal codec to Decode it.

**Comment by smarterclayton** [comment]
Yeah I think we will cache these in our clients (disk and web local
storage) and refresh them periodically.

**Comment by lavalamp** [comment]
Even when caching things, you know what you're caching, it's not a mystery file that you need to do introspection on, so I don't think that makes it urgent, either.

OK, I chatted with @bgrant0607 about this, and we came up with:
* If it's as easy as just adding a Kind field to the top level objects, then we can go ahead and do that (esp. if @liggitt or @deads2k wants to make a PR ;) )
* If it's a big hairball where we have to figure out how to get a TypeMeta in there to make the encoder happy while not emitting an apiVersion field, we just don't have time to do that for 1.1.

**Comment by liggitt** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/16364 to add it

**Comment by lavalamp** [comment]
Thanks @liggitt. @caesarxuchao can you rebase this on top of that PR, and omit the JSON codec?

**Comment by caesarxuchao** [comment]
Sure.

**Comment by liggitt** [comment]
jsonCodec (or straight JSON unmarshal) is needed against 1.0 servers (so /api) but can be omitted for resource list, group list and group queries. By the raw JSON decode, add a comment that that is only for 1.0 compatibility and can be removed once support for 1.0 is dropped

**Comment by caesarxuchao** [comment]
@liggitt actually we can omit jsonCodec completely. If you can recall the error I got was that the decoder couldn't find object in schema, and its the client code who supplied the object (unversioned.APIVersions). Now that APIVersions is added to the Schema in your PR, the error is gone.

So actually we can get rid of the jsonCodec by adding discovery APIs to Schema, without adding TypeMeta to the types..

File: pkg/client/unversioned/discovery_client.go [hunk 14]
```diff
@@ -155,6 +141,8 @@ func (d *DiscoveryClient) ServerResources() (map[string]*unversioned.APIResource
...
+	// Discovery client deals with unversioned objects, so we use api.Codec.
+	config.Codec = api.Codec
```
Comment (caesarxuchao) [inline pkg/client/unversioned/discovery_client.go]: changed to api.Codec

File: pkg/conversion/decode.go [hunk 15]
```diff
@@ -19,6 +19,7 @@ package conversion
 import (
 	"errors"
 	"fmt"
+
```
Comment (caesarxuchao) [inline pkg/conversion/decode.go]: This is not necessary, but the change is consistent with other files, so I kept this change.

**Comment by caesarxuchao** [comment]
@lavalamp @liggitt updated based on #16364. PTAL. Thank you.

I will squash the commits before merge.

**Comment by bgrant0607** [comment]
#16364 has been LGTMed. Thanks.

**Comment by lavalamp** [comment]
Looks good, will put on the label when #16364 merges.

**Comment by lavalamp** [comment]
@caesarxuchao can you make an issue for this, so we can track it for 1.1? Thanks

**Comment by lavalamp** [comment]
Oh wait, you're occupied today-- I'll make one.

**Comment by caesarxuchao** [comment]
@lavalamp rebased and squashed. I'll make a cherry-pick now.

**Comment by caesarxuchao** [comment]
Adding the lgtm label per https://github.com/kubernetes/kubernetes/pull/16082#issuecomment-151681967. @lavalamp please remove it if you want to.

**Comment by lavalamp** [comment]
That's fine, thanks!

**Comment by lavalamp** [comment]
@k8s-bot unit test this
